plutus-core/untyped-plutus-core/test/Transform/Inline/Spec.hs

```diff
@@ -0,0 +1,100 @@
+import Numeric.Natural (Natural)
+import PlutusCore.Annotation (Inline (MayInline))
+import PlutusCore.Default (DefaultFun (AddInteger))
+import PlutusCore.Quote (runQuote)
+import PlutusCore.Size (Size (..))
+import PlutusPrelude (def)
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.HUnit (assertBool, testCase)
+import UntypedPlutusCore (DefaultUni, Name (..), Term (..), Unique (..))
+import UntypedPlutusCore.Transform.Inline (InlineHints (InlineHints), InlineInfo (..), InlineM,
+                                           S (..), Subst (Subst), TermEnv (TermEnv),
+                                           isFirstVarBeforeEffects)
+
+--------------------------------------------------------------------------------
+-- Tests -----------------------------------------------------------------------
+
+test_inline :: TestTree
+test_inline =
+  testGroup
+    "isFirstVarBeforeEffects"
+    [ testCase "before effects" testVarBeforeEffects
+    , testCase "after effects" testVarAfterEffects
+    ]
+
+testVarBeforeEffects :: IO ()
+testVarBeforeEffects = do
+  assertBool "var1 is evaluated before effects" do
+    testFirstVarBeforeEffects (varName 1) term
+  assertBool "var2 is evaluated before effects" do
+    testFirstVarBeforeEffects (varName 2) term
+
+testVarAfterEffects :: IO ()
+testVarAfterEffects = do
+  assertBool "var3 is not evaluated after effects" $ not do
+    testFirstVarBeforeEffects (varName 3) term
+
+--------------------------------------------------------------------------------
+-- Test term: ------------------------------------------------------------------
+
+term :: Term Name DefaultUni DefaultFun ()
+term =
+  {- Evaluation order:
+
+    1. pure work-free: var1
+    2. pure work-free: var2
+    3. impure? maybe work?: addInteger var1 var2
+    4. pure work-free: var3
+    5. impure? maybe work?: addInteger (addInteger var1 var2) var3
+  -}
+  add `ap` (add `ap` varTerm 1 `ap` varTerm 2) `ap` varTerm 3
+
+--------------------------------------------------------------------------------
+-- Helper functions: -----------------------------------------------------------
+
+testFirstVarBeforeEffects :: Name -> Term Name DefaultUni DefaultFun () -> Bool
+testFirstVarBeforeEffects name = runInlineM . isFirstVarBeforeEffects name
+
+runInlineM :: InlineM Name DefaultUni DefaultFun () r -> r
+runInlineM m = result
+ where
+  (result, _finalState) =
+    runQuote (runStateT (runReaderT m inlineInfo) initialState)
+
+  inlineInfo :: InlineInfo Name DefaultFun ()
+  inlineInfo =
+    InlineInfo
+      { _iiUsages = MultiSet.empty
+      , _iiHints = InlineHints \_ann _name -> MayInline
+      , _iiBuiltinSemanticsVariant = def
+      , _iiInlineConstants = True
+      , _iiInlineCallsiteGrowth = Size 1_000_000
+      }
+
+  initialState :: S Name DefaultUni DefaultFun ()
+  initialState = S{_subst = Subst (TermEnv mempty), _vars = mempty}
+
+-- Term construction: ----------------------------------------------------------
```

Comment:
Instead of placing the term construction functions here, I suggest taking this opportunity to organize the test code better. There are many test cases in UPLC tests that involve constructing terms - see `Transform.Simplify` and `Transform.CaseOfCase.Test`. If you are going to add these helper functions, put them in a test library so that it can be used in all tests.

Also, make the module names consistent - e.g. `Transform.CaseOfCase.Test` vs. `Transform.Inline.Spec`.

Reply:
I thought about making a test-reusable term building facility and using it in all tests. I decided to postpone it to other times for a few reasons: 1. not to broaden the scope of change for this particular task. 2. to accumulate some term-building code here and there before coming up with an abstraction. This way abstractions usually come out more adequate, as they are well informed by the accumulated usages.

Reply:
I made another PR with a general UPLC term construction functions that are general enough to be useful in other modulel (I demonstrated this usefulness by applying those helpers in the `Transform.Simplify` spec). This way term construction helpers from this module are replaced with the helpers from the `UntypedPlutusCore.Test.Term.Construction`.

Reply:
In which sense to you think that `mkIterAppNoAnn` is "better"?
```haskell
term = mkIterAppNoAnn (var f) [var a, var b, var c]
```
vs. 
```haskell
term = var f `app` var a `app` var b `app` var c
```
I also don't see how is it more general: I can't imagine anything that `mkIterAppNoAnn` can express that `ann` can't.

Reply:
Regardless of which is better - a typical way of introducing tech debt is to duplicate some functionality that's already there, such that two slightly different APIs co-exist, and different code use different ones.

Maybe yours is slightly better (which is questionable - for example, one way `mkIterAppNoAnn` is better is that it is more similar to the GHC API for constructing GHC Core terms - consistency is good), but I think the right approach is (1) use the exiting api for this PR (2) open an issue proposing a new API (3) let's review and agree on it, before implementing the new API and retiring the old one.